src/lr_scheduler/cosine_annealing_lr.rs

```diff
@@ -0,0 +1,51 @@
+use super::Schedule;
+
+pub struct CosineAnnealingLR {
+    pub base_lr: f64,
+    pub t_max: usize,
+    pub eta_min: f64,
+}
+
+impl Default for CosineAnnealingLR {
+    fn default() -> Self {
+        CosineAnnealingLR {
+            base_lr: 0.05,
+            t_max: 5,
+            eta_min: 0.0,
+        }
+    }
+}
+
+impl Schedule for CosineAnnealingLR {
+    fn get_lr(&self, lr: f64, last_epoch: usize) -> f64 {
+        if last_epoch == 0 {
+            return lr;
+        }
+
+        let pi = std::f64::consts::PI;
+        let lr_range = self.base_lr - self.eta_min;
+        let wavelen_percent = pi * (last_epoch as f64) / (self.t_max as f64);
+
+        if last_epoch == 1 {
```

Comment:
Pytorch has this logic under the condition `elif self._step_count == 1 and self.last_epoch > 0:` is this accurate?

Reply:
Since I did away with `step_count`, it's hard to say with absolute certainty. I would say _probably_, considering that this is in `get_lr` and not `get_closed_form_lr`. Within `get_lr`, PyTorch's `step_count` and our `last_epoch` can be treated as one and the same.

Reply:
We should, however, introduce `_step_count` in the next PR involving LR schedulers, as it seems to get more use when multiple schedulers for one model exist.